src/rulesets/Base/mapreduce.jl

```diff
@@ -417,137 +417,184 @@ end
+            # Don't need to store every `da`, need one for the next iteration + the last.
         end
         dop = sum(first, trio)
-        dx = map(last, _reverse1(trio))
-        if init === _InitialValue()
-            # `hobbits` is one short
-            dx = _vcat1(trio[end][2], dx)
-        end
-        return (NoTangent(), dop, project(_reshape1(dx, axe)))
+        dx = (trio[end][2], reverse(map(last, trio))...)
+        return (NoTangent(), NoTangent(), ProjectTo(op)(dop), NoTangent(), project(dx))
     end
-    return y, unfoldl
+    return y, foldl_pullback_tuple
 end
 
+function rrule(
+        config::RuleConfig{>:HasReverseMode},
+        ::typeof(Base.mapfoldl_impl),
+        ::typeof(identity),
+        op::G, 
+        init, 
+        x::Tuple;
+    ) where {G}
+    # Treat `init` by simply appending it to the `x`:
+    y, back = rrule(config, Base.mapfoldl_impl, identity, op, Base._InitialValue(), (init, x...))
+    project_x = ProjectTo(x)
+    project_in = ProjectTo(init)
+    function foldl_pullback_tuple_init(dy)
+        _, _, dop, _, dxplus = back(dy)
+        return (NoTangent(), NoTangent(), dop, project_in(first(dxplus)), project_x(Base.tail(dxplus)))
+    end
+    return y, foldl_pullback_tuple_init
+end
 
 #####
-##### Iterator-or-Tuple functions
+##### `foldl(f, ::Array)`
 #####
 
-# This zoo of underscore functions helps `foldl` & `accumulate` handle both tuples and arrays,
-# and also provides some alternatives for versions of Julia where iterators weren't supported.
-# Inspired by `Base._reverse`, used in defn of `foldr`.
-
-# To support 2nd derivatives, some may need their own gradient rules. And _drop1 should perhaps
-# be replaced by _peel1 like Iterators.peel
+# The implementation was originally for both tuples and arrays, although using accumulate
+# to carry intermediate results along creates arrays of tuples which could be avoided.
+# Using a loop can be a few times faster, this should be replaced:
+# https://github.com/FluxML/Zygote.jl/issues/644#issuecomment-628762305
 
-_reverse1(x) = Iterators.reverse(x)
-_drop1(x) = Iterators.drop(x, 1)
-_zip2(x, y) = zip(x, y)  # for `accumulate`, below
+# Note also that it does not return a gradient for `init`, now marked `@not_implemented`.
 
-_reverse1(x::Tuple) = reverse(x)
-_drop1(x::Tuple) = Base.tail(x)
-_zip2(x::Tuple{Vararg{Any,N}}, y::Tuple{Vararg{Any,N}}) where N = ntuple(i -> (x[i],y[i]), N)
-
-struct _InitialValue end  # Old versions don't have `Base._InitialValue`
+function rrule(
+        config::RuleConfig{>:HasReverseMode}, ::typeof(Base.mapfoldl_impl), ::typeof(identity), op::G, init, x::Union{AbstractArray, Tuple};
+    ) where {G}
+    start, list = if init === Base._InitialValue()
+        Iterators.peel(x)
+    else
+        # Case with init keyword is simpler to understand first!
+        init, x
+    end
+    hobbits = accumulate(list; init=(start, nothing)) do (a, _), b
+        c, back = rrule_via_ad(config, op, a, b)
```

Comment:
Is there any way to not capture the accumulated outputs (`c`s) in the pullback? It seems easy enough for tuples using `map`, but I'm unsure if the extra allocation would be welcomed for arrays.

Reply:
Yes, you can write a `for` loop like this: https://github.com/FluxML/Zygote.jl/issues/644#issuecomment-628762305 . IMO this array method should probably be replaced, but not today. 

Carrying `c` by updating a variable from inside `accumulate` was very slow, IIRC it hits the closure issue.

Reply:
I actually tried that in https://github.com/FluxML/Flux.jl/pull/2003. The main challenges are nested differentiation and handling the case when `typeof(x |> f) != typeof(x |> f |> f)` (you must widen, which means preallocating an array is impossible without `return_type` shenanigans).

So, assuming type inference cooperates, the accumulate approach seems no less promising. Would there be any objections to a post-processing step like the following which allows the GC to clean up intermediate outputs before the pullback?

```julia
# ... y = first(last(hobbits))
# If outputs are allocated inline, we're less worried about memory overhead
# and the GC can't free them individually anyways. 
if !isbitstype(eltype(hobbits))
  hobbits = map(((_, pb)) -> (nothing, pb),  hobbits)
end
# axe = axes(x) ...
```

Reply:
Yes the mutation has these problems, but was much quicker, maybe it can be used when safe. 

The intention with writing `foldl` in terms of `accumulate` was to allow for 2nd derivatives, but not sure this actually works right now. 

Re saving memory, we can add something like `unzip_accumulate(f, xs; init) = StructArrays.components(StructArray(Iterators.accumulate(f, xs; init)))` to free the bits we don't need anymore. 
```
julia> accumulate([1,2,3], init=(4,5)) do prev, this
         this .+ prev
       end
3-element Vector{Tuple{Int64, Int64}}:
 (5, 6)
 (7, 8)
 (10, 11)

julia> unzip_accumulate([1,2,3], init=(4,5)) do prev, this
         this .+ prev
       end
([5, 7, 10], [6, 8, 11])
```

But this PR would like to kick the can down the road on such improvements. 

(And others -- it returns `@not_implemented` for `accumulate`'s init, might be easy to do better, but tired of adding tests... at least it's no longer wrong.)